### PR TITLE
Serverprofile

### PR DESCRIPTION
**Pull request by hagna**
This adds the serverprofile option.

endSrvProfile could also be called from a finalizer for Server if you want.

### CODE REVIEW

**Comment by rminnich** [comment]
I'm lost -- when did the finalizer get added? I better get back up to speed with ninep :-)

you didn't add that trace change I hope from the earlier PR.

File: stub/stub.go [hunk 1]
```diff
@@ -8,13 +8,20 @@ package stub
...
+	"io/ioutil"
 	"log"
+	"os"
+	"path/filepath"
 	"runtime"
+	"runtime/pprof"
 	"sync/atomic"
 )
 
+var serverprofile = flag.String("serverprofile", "", "This is for specifying the prefix of the output file for the profile")
```
Comment (sevki) [inline stub/stub.go]: I really dislike this paradigm defining flags that are not in or around the main function. I think it misses the point of having readable code, and the go community at large seems to agree. 

Could we move this inside Server struct?

> Only func main has the right to decide which flags are available to the user. 

[source](https://peter.bourgon.org/go-best-practices-2016/#top-tip-5)

Comment (rminnich) [inline stub/stub.go]: That's how Go is written. There's all kinds of packages I use that define their own flags.

File: stub/stub.go [hunk 2]
```diff
@@ -539,12 +563,15 @@ func (s *Server) readNetPackets() {
 	if s.Trace != nil {
 		s.Trace("Starting readNetPackets")
 	}
+	if *serverprofile != "" {
+		s.beginSrvProfile()
```
Comment (rminnich) [inline stub/stub.go]: After this line
defer func() {
s.endSrvProfile()
}

Leave the returns as is below i.e. don't change them to break

File: stub/stub.go [hunk 3]
```diff
@@ -570,12 +597,15 @@ func (s *Server) readNetPackets() {
...
 			log.Printf("readNetPackets: write error: %v", err)
 			s.Dead = true
-			return
+			break
 		}
 		if s.Trace != nil {
 			s.Trace("Returned %v amt %v", b, amt)
 		}
 	}
+	if *serverprofile != "" {
```
Comment (rminnich) [inline stub/stub.go]: Remove these three lines because the are in the defer above.

File: stub/stub.go (hunk 2 above)
Comment (rminnich) [inline stub/stub.go]: Sorry

defer s.endSrvProfile()

sorry.

**Comment by rminnich** [comment]
See my note on user of defer, but this is close. Thanks.

**Comment by hagna** [comment]
I'm curious why return instead

File: stub/stub.go [hunk 4]
```diff
@@ -565,13 +565,14 @@ func (s *Server) readNetPackets() {
 	}
 	if *serverprofile != "" {
 		s.beginSrvProfile()
+        defer s.endSrvProfile()
```
Comment (rminnich) [inline stub/stub.go]: go fmt -w on this file. The indent looks wrong.

**Comment by rminnich** [comment]
The main reason for return in this case is that are done and want to make sure nothing else happens. The only reason for a break would be to do exit processing, but you've covered that with the defer. The combination of return and defer is a lot more reliable than break and exit processing.

I think you need to gofmt  -w this file and then we're good to go. Thanks.

**Comment by rminnich** [comment]
LGTM

**Pull request merged**